### PR TITLE
[SPARK-37592][SQL] Improve performance of `JoinSelection`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
When I reading the implement of AQE, I find the process select join with hint exists a lot cumbersome code.

The join hint has a relatively high learning curve for users, so the SQL not  contains join hint in more cases. 

### Why are the changes needed?
Improve performance of `JoinSelection`

### Does this PR introduce _any_ user-facing change?
'No'.
Just change the inner implement.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala [hunk 1]
```diff
@@ -548,11 +548,9 @@ case class ApplyColumnarRulesAndInsertTransitions(
 
   def apply(plan: SparkPlan): SparkPlan = {
     var preInsertPlan: SparkPlan = plan
-    columnarRules.foreach((r : ColumnarRule) =>
-      preInsertPlan = r.preColumnarTransitions(preInsertPlan))
+    columnarRules.foreach( r => preInsertPlan = r.preColumnarTransitions(preInsertPlan))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:551]: how is this related to `JoinSelection`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -266,11 +266,15 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
             }
         }
 
-        createBroadcastHashJoin(true)
-          .orElse { if (hintToSortMergeJoin(hint)) createSortMergeJoin() else None }
-          .orElse(createShuffleHashJoin(true))
-          .orElse { if (hintToShuffleReplicateNL(hint)) createCartesianProduct() else None }
-          .getOrElse(createJoinWithoutHint())
+        if (hint.isEmpty) {
+          createJoinWithoutHint()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:270]: this change LGTM

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:270]: Can we do this in `case logical.Join(left, right, joinType, condition, hint) ...` as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:551]: Just simplify the code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:270]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:551]:
```suggestion
    columnarRules.foreach(r => preInsertPlan = r.preColumnarTransitions(preInsertPlan))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala [hunk 3]
```diff
@@ -548,11 +548,9 @@ case class ApplyColumnarRulesAndInsertTransitions(
...
     var postInsertPlan = insertTransitions(preInsertPlan, outputsColumnar)
-    columnarRules.reverse.foreach((r : ColumnarRule) =>
-      postInsertPlan = r.postColumnarTransitions(postInsertPlan))
+    columnarRules.reverse.foreach( r => postInsertPlan = r.postColumnarTransitions(postInsertPlan))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:553]:
```suggestion
    columnarRules.reverse.foreach(r => postInsertPlan = r.postColumnarTransitions(postInsertPlan))
```

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks a lot!

**Comment by dongjoon-hyun** [comment]
+1, LGTM.
